x-pack/plugins/files/public/components/upload_file/upload_file.tsx

```diff
@@ -0,0 +1,165 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { EuiFilePicker } from '@elastic/eui';
+import React, { type FunctionComponent, useState, useRef, useCallback } from 'react';
+import useObservable from 'react-use/lib/useObservable';
+import { finalize } from 'rxjs';
+import { FileKind } from '../../../common';
+import { FilesClient } from '../../types';
+
+import { useFilesContext } from '../context';
+import { useBehaviorSubject } from '../use_behavior_subject';
+
+import { UploadFile as Component } from './upload_file.component';
+import { createUploadState } from './upload_state';
+
+/**
+ * An object representing an uploadded file
+ */
+interface UploadedFile {
+  /**
+   * The ID that was generated for the uploaded file
+   */
+  id: string;
+  /**
+   * The kind of the file that was passed in to this component
+   */
+  kind: string;
+}
+
+/**
+ * UploadFile component props
+ */
+export interface Props<Kind extends string = string> {
+  /**
+   * A file kind that should be registered during plugin startup. See {@link FileServiceStart}.
+   */
+  kind: Kind;
+  /**
+   * A files client that will be used process uploads.
+   */
+  client: FilesClient;
+  /**
+   * Allow users to clear a file after uploading.
+   *
+   * @note this will NOT delete an uploaded file.
+   */
+  allowClear?: boolean;
+  /**
+   * Start uploading the file as soon as it is provided
+   * by the user.
+   */
+  immediate?: boolean;
+  /**
+   * Metadata that you want to associate with any uploaded files
+   */
+  meta?: Record<string, unknown>;
+  /**
+   * Whether this component should display a "done" state after processing an
+   * upload or return to the initial state to allow for another upload.
+   *
+   * @default false
+   */
+  allowRepeatedUploads?: boolean;
+  /**
+   * Called when the an upload process fully completes
+   */
+  onDone: (files: UploadedFile[]) => void;
+
+  /**
+   * Called when an error occurs during upload
+   */
+  onError?: (e: Error) => void;
+}
+
+/**
+ * This component is intended as a wrapper around EuiFilePicker with some opinions
+ * about upload UX. It is optimised for use in modals, flyouts or forms.
+ *
+ * In order to use this component you must register your file kind with {@link FileKindsRegistry}
+ */
+export const UploadFile: FunctionComponent<Props> = ({
+  meta,
+  client,
+  onDone,
+  onError,
+  allowClear,
+  kind: kindId,
+  immediate = false,
+  allowRepeatedUploads = false,
+}) => {
+  const { registry } = useFilesContext();
+
+  const ref = useRef<null | EuiFilePicker>(null);
+
+  const [kind] = useState<FileKind>(() => registry.get(kindId));
+  const [uploadState] = useState(() =>
+    createUploadState({
+      client,
+      fileKind: kind,
+    })
+  );
```

Comment:
Now you have 3 entities that represent your state: (1) `client`; (2) `{ registry }`; (3) and `uploadState`. And only the `{ registry }` is available through React context. Would be nice to combine them into one thing and provide that one thing though React context:

```suggestion
export const UploadFile: FunctionComponent<Props> = ({
  client,
  kind,
  ...rest,
}) => {
  const { registry } = useFilesContext();
  const uploadState = useMemo(() => new UploadState({
    client,
    registry,
    kind,
  }), [client, kind, registry]);


  return (
    <context.Provider value={uploadState}>
      <UploadFileConnected {...rest} />
    <context.Provider>
  );
```

, where

```tsx
export const context = React.createContext<UploadState | null>(null);
export const useUploadState = React.useContext(context)!;
```